### PR TITLE
Support per-backend role and destination in replication engine

### PR DESCRIPTION
**Pull request by maeldonn**
Read destination bucket and role from each backends[] entry instead of the shared top-level fields, so a single source object can be replicated to multiple CRR destinations with their own role. Legacy entries without per-backend fields keep working via top-level fallback in ObjectQueueEntry's site-aware getters; MongoQueueProcessor's oplog path now matches every applicable rule and dedups backends per the design's (site, destination, role) rule.

Issue: BB-762

### CODE REVIEW

File: extensions/replication/tasks/ReplicateObject.js [hunk 1]
```diff
@@ -287,18 +289,36 @@ class ReplicateObject extends BackbeatTask {
...
                 return cb(errors.BadRole);
             }
-            if (roles[1] !== entryRoles[1]) {
+            // Multi-destination CRR: derive the expected destination
+            // role for this site from the matching rule's Account
+            // override; legacy configs without Account fall back to
+            // the literal two-comma role equality check.
+            const matchingRule = data.ReplicationConfiguration.Rules.find(
+                rule => rule.Status === 'Enabled' &&
+                    entry.getObjectKey().startsWith(rule.Prefix) &&
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/replication/tasks/ReplicateObject.js]: Updated.

[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/replication/tasks/ReplicateObject.js]: Updated.

File: extensions/replication/queueProcessor/QueueProcessor.js [hunk 2]
```diff
@@ -864,41 +863,75 @@ class QueueProcessor extends EventEmitter {
...
+        const pendingBackends = sourceEntry.getReplicationBackends()
+            .filter(b => b.status === 'PENDING' && b.site === this.site);
+
+        if (pendingBackends.length === 0) {
+              logSkip();
+              return;
+        }
+
+        const endpointType = this.destConfig.replicationEndpoint?.type;
+        const isCloud = endpointType && this.replicationBackends?.includes(endpointType);
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/replication/queueProcessor/QueueProcessor.js]: Updated.

File: extensions/replication/tasks/ReplicateObject.js [hunk 3]
```diff
@@ -287,18 +299,35 @@ class ReplicateObject extends BackbeatTask {
...
+                    rule.Destination &&
+                    rule.Destination.StorageClass === this.site);
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/replication/tasks/ReplicateObject.js]: Fixed: the rule lookup now also matches `Destination.Account` against the account ID embedded in the entry's per-backend role, so each fan-out task picks the rule that produced its own role (not whichever rule `.find()` returns first).

File: extensions/replication/tasks/ReplicateObject.js [hunk 4]
```diff
@@ -266,9 +267,9 @@ class ReplicateObject extends BackbeatTask {
                     'replication disabled for object'));
             }
             const roles = data.ReplicationConfiguration.Role.split(',');
-            if (roles.length !== 2) {
-                log.error('expecting two roles separated by a ' +
-                    'comma in bucket replication configuration',
+            if (roles.length < 1 || roles.length > 2) {
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/replication/tasks/ReplicateObject.js]: Done.

File: package.json [hunk 5]
```diff
@@ -54,7 +54,7 @@
     "@scality/cloudserverclient": "^1.0.8",
     "@smithy/node-http-handler": "^3.3.3",
     "JSONStream": "^1.3.5",
-    "arsenal": "git+https://github.com/scality/arsenal#8.3.9",
+    "arsenal": "git+https://github.com/scality/arsenal#21b9bb33ad77d21609a690a5709a645eab1a95d7",
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply package.json]: Arsenal is not released yet.

File: package.json [hunk 6]
```diff
@@ -54,7 +54,7 @@
     "@scality/cloudserverclient": "^1.0.8",
     "@smithy/node-http-handler": "^3.3.3",
     "JSONStream": "^1.3.5",
-    "arsenal": "git+https://github.com/scality/arsenal#8.3.9",
+    "arsenal": "git+https://github.com/scality/arsenal#39c1a642d77436d871305435dcf244a64423a4b8",
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply package.json]: Arsenal is not released yet.

[the remark this answers is not in the document]
Comment (maeldonn) [reply package.json]: Arsenal is not released yet.

File: package.json [hunk 7]
```diff
@@ -54,7 +54,7 @@
     "@scality/cloudserverclient": "^1.0.8",
     "@smithy/node-http-handler": "^3.3.3",
     "JSONStream": "^1.3.5",
-    "arsenal": "git+https://github.com/scality/arsenal#8.3.9",
+    "arsenal": "git+https://github.com/scality/arsenal#300556545ef6f8350c27a9501f66c2b7d5d36a0d",
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply package.json]: Arsenal is not released yet.

[the remark this answers is not in the document]
Comment (maeldonn) [reply package.json]: Arsenal is not released yet.

File: package.json [hunk 8]
```diff
@@ -54,7 +54,7 @@
     "@scality/cloudserverclient": "^1.0.8",
     "@smithy/node-http-handler": "^3.3.3",
     "JSONStream": "^1.3.5",
-    "arsenal": "git+https://github.com/scality/arsenal#8.3.9",
+    "arsenal": "git+https://github.com/scality/arsenal#e4860ebc364db6c3cd5b3f3210339d0fa512ed24",
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply package.json]: Arsenal is not released yet.

File: package.json [hunk 9]
```diff
@@ -54,7 +54,7 @@
     "@scality/cloudserverclient": "^1.0.8",
     "@smithy/node-http-handler": "^3.3.3",
     "JSONStream": "^1.3.5",
-    "arsenal": "git+https://github.com/scality/arsenal#8.3.9",
+    "arsenal": "git+https://github.com/scality/arsenal#c3f3bffde6eeb935e03847df010ac37ac7ba8944",
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply package.json]: arsenal is not released yet

File: extensions/replication/queueProcessor/QueueProcessor.js [hunk 10]
```diff
@@ -864,41 +863,75 @@ class QueueProcessor extends EventEmitter {
...
+                role: backend.role,
+            });
+
+            await new Promise((resolve, reject) => {
+                this.taskScheduler.push(
+                    { task, entry: perBackendEntry, kafkaEntry },
+                    err => err ? reject(err) : resolve(),
+                );
+            });
         }
```
[the remark this answers is not in the document]
Comment (SylvainSenechal) [reply extensions/replication/queueProcessor/QueueProcessor.js:952]: Good raised question ?

File: extensions/replication/tasks/ReplicateObject.js [hunk 11]
```diff
@@ -254,7 +255,8 @@ class ReplicateObject extends BackbeatTask {
             .then(data => {
                 const replicationEnabled = (
                 data.ReplicationConfiguration.Rules.some(
-                    rule => entry.getObjectKey().startsWith(rule.Prefix)
+                    rule => entry.getObjectKey().startsWith(
```
Comment (SylvainSenechal) [inline extensions/replication/tasks/ReplicateObject.js:258]: Is this a change of behavior with older version because of the '' fallback ?

edit: Ok i guess startWith('') is always true, so I guess its fine unless we want to treat an empty prefix filter as invalid ?

File: extensions/replication/tasks/ReplicateObject.js [hunk 12]
```diff
@@ -266,9 +268,9 @@ class ReplicateObject extends BackbeatTask {
                     'replication disabled for object'));
             }
             const roles = data.ReplicationConfiguration.Role.split(',');
-            if (roles.length !== 2) {
-                log.error('expecting two roles separated by a ' +
-                    'comma in bucket replication configuration',
+            if (roles.length > 2) {
```
Comment (SylvainSenechal) [inline extensions/replication/tasks/ReplicateObject.js:271]: nit: check len == 1 || 2 ? 
But also, are you sure we shouldn't just enforce === 2 ? In what case could it be 1

File: package.json [hunk 13]
```diff
@@ -54,7 +54,7 @@
     "@scality/cloudserverclient": "^1.0.8",
     "@smithy/node-http-handler": "^3.3.3",
     "JSONStream": "^1.3.5",
-    "arsenal": "git+https://github.com/scality/arsenal#8.3.9",
+    "arsenal": "git+https://github.com/scality/arsenal#724d06289e11a429dfdf9e1236f98e904f748a90",
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply package.json]: arsenal is not yet released.

File: package.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (maeldonn) [reply package.json]: arsenal is not yet released.

File: package.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (maeldonn) [reply package.json]: arsenal is not yet released.

[the remark this answers is not in the document]
Comment (maeldonn) [reply package.json]: arsenal is not yet released.

File: extensions/mongoProcessor/MongoQueueProcessor.js [hunk 14]
```diff
@@ -340,53 +322,33 @@ class MongoQueueProcessor {
...
+
+        if (backends.length === 0) {
+            return;
         }
+
+        entry.setReplicationInfo({
+            status: 'PENDING',
+            backends,
+            content,
+            role: ReplicationConfiguration.resolveSourceRole(bucketRepInfo.role),
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/mongoProcessor/MongoQueueProcessor.js:349]: Confirmed — grep over the production code shows no consumer reads `replicationInfo.destination` or `replicationInfo.storageClass` at the top level. Per-backend destination/storageClass are now carried inside `backends[]`, set by `ReplicationConfiguration.resolveBackends`.

File: extensions/replication/queueProcessor/QueueProcessor.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/replication/queueProcessor/QueueProcessor.js:952]: Switched to per-backend try/catch: each pending backend is attempted independently this pass, and the first error is re-thrown at the end so the Kafka entry is retried. The `status === 'PENDING'` filter ensures only the still-unfinished backends are re-attempted on retry.

File: lib/models/ObjectQueueEntry.js [hunk 15]
```diff
@@ -179,58 +205,60 @@ class ObjectQueueEntry extends ObjectMD {
...
+            .setReplicationSiteStatus(key, 'PENDING')
             .setReplicationStatus(this._getGlobalReplicationStatus())
             .setOriginOp('s3:Replication:OperationPendingReplication');
     }
 
-    toRetryEntry(site) {
+    toRetryEntry(key) {
+        const backend = this._findBackend(key);
         return this.clone()
             .setAccountId(this.getAccountId())
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply lib/models/ObjectQueueEntry.js:270]: Confirmed — `_findBackend(key)` is defined on `ObjectMD` in the pinned arsenal (`node_modules/arsenal/build/lib/models/ObjectMD.js:973`, also declared in the `.d.ts`).

File: extensions/replication/tasks/ReplicateObject.js (hunk 11 above)
Comment (maeldonn) [reply extensions/replication/tasks/ReplicateObject.js:258]: Right — `startsWith('')` is always true, which matches S3 semantics for a rule with no prefix (matches all keys). The previous code only checked `rule.Prefix`, so V2 rules using `Filter.Prefix` were silently skipped. An empty prefix is valid per the S3 spec, so we don't want to treat it as invalid.

File: extensions/replication/tasks/ReplicateObject.js (hunk 12 above)
Comment (maeldonn) [reply extensions/replication/tasks/ReplicateObject.js:271]: Length 1 is a valid legacy case — arsenal's `ReplicationConfiguration.resolveDestinationRole` documents that the top-level `Role` accepts either a single ARN (used for both source and destination) or a `source,destination` pair. Enforcing `=== 2` would reject single-ARN configs, so the check stays as `> 2` (reject only excess segments).

File: package.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (maeldonn) [reply package.json]: arsenal is not yet released.

File: lib/models/ObjectQueueEntry.js [hunk 16]
```diff
@@ -179,58 +230,60 @@ class ObjectQueueEntry extends ObjectMD {
...
                      bucket: this.getBucket(),
                      key: this.getObjectVersionedKey(),
                      value: JSON.stringify(
                          this.getIgnore() ? {} : this.getValue()),
-                     site,
+                     site: backend?.site,
+                     destination: backend?.destination,
+                     role: backend?.role,
                      workflowId: this.getWorkflowId(),
                      workflowVersion: this.getWorkflowVersion(),
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply lib/models/ObjectQueueEntry.js:288]: Old entries still work. The consumer filter only reads `status` and `site`, which existed before. `getReplicationRoles` falls back to the top-level role when no per-backend role is set, and `backend.destination` is only used in a log line; the real destination comes from `destConfig`. So old entries keep their legacy behavior, and new entries get the per-backend overrides.

File: extensions/mongoProcessor/MongoQueueProcessor.js [hunk 17]
```diff
@@ -340,53 +322,33 @@ class MongoQueueProcessor {
...
+            entry.getObjectKey(),
+            isCloud,
+            zenkoObjMd?.replicationInfo?.backends,
+        );
+
+        if (backends.length === 0) {
+            return;
         }
+
+        entry.setReplicationInfo({
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/mongoProcessor/MongoQueueProcessor.js:345]: Backbeat is deployed as a single versioned unit in S3C/Artesca, so readers and writers upgrade together; no mixed-version window in production. The reverse direction (new reader, old data) is handled by arsenal getter fallbacks, see the sibling thread on `ObjectQueueEntry.toKafkaEntry`.

**Review by delthas — APPROVED** [review]
LGTM on the surface, but def need someone with more hands-on exp with CRR to review in depth.

Also I found it hard to review (+2600, -600 on 20~ files), and the commit split did not help much. Wondering if perhaps future MRs could have smaller, atomic commits. Here there's a big commit and a test commit and then 2 review fixups commits (which I personally find make it even harder to review cause we have to go "oh there's an issue here" remember then "oh it's fixed in that upcoming commit"  but we acted in retro that we want no squashes so OK).

File: lib/models/ObjectQueueEntry.js [hunk 18]
```diff
@@ -179,58 +230,60 @@ class ObjectQueueEntry extends ObjectMD {
...
-    toRetryEntry(site) {
+    toRetryEntry(backend) {
+        const matched = this._findBackend(backend);
         return this.clone()
             .setAccountId(this.getAccountId())
-            .setReplicationBackends(this.getReplicationBackends().filter(o => o.site === site))
-            .setReplicationSiteStatus(site, 'PENDING')
-            .setReplicationStorageClass(site)
+            .setReplicationBackends(matched ? [matched] : [])
+            .setReplicationSiteStatus(backend, 'PENDING')
```
Comment (SylvainSenechal) [inline lib/models/ObjectQueueEntry.js:272]: no more setReplicationStorageClass ?

Comment (SylvainSenechal) [reply lib/models/ObjectQueueEntry.js:272]: no i think its ok

File: tests/unit/lib/models/ObjectQueueEntry.spec.js [hunk 19]
```diff
@@ -1,25 +1,355 @@
...
+                {
+                    site: 'siteB',
+                    status: 'PENDING',
+                    dataStoreVersionId: '',
+                    destination: 'arn:aws:s3:::bucket-b',
+                    role: 'arn:aws:iam::333:role/dst',
+                },
+            ]);
+
+            assert.strictEqual(entry.getReplicationTargetBucket({ site: 'siteA' }), 'bucket-a');
```
Comment (SylvainSenechal) [inline tests/unit/lib/models/ObjectQueueEntry.spec.js:62]: Looking at this, I have a question not directly related to this test, but looking at line 43 (destination: 'arn:aws:s3:::bucket-a',) made me think about it :
 
Thomas recently fixed a bug in Arsenal, where an ARN was miss parsed because there were ':' directly in the object keys name https://github.com/scality/Arsenal/pull/2638/changes

I think here its ok as we are working on bucket names, but I think just nice to keep this in mind as these pr are working on ARNs

File: tests/unit/lib/models/ObjectQueueEntry.spec.js [hunk 20]
```diff
@@ -1,25 +1,355 @@
...
+            assert.strictEqual(entry.getReplicationTargetBucket({ site: 'siteB' }), 'bucket-b');
+        });
+
+        it('falls back to top-level role string when backend has none', () => {
```
Comment (SylvainSenechal) [inline tests/unit/lib/models/ObjectQueueEntry.spec.js:66]: Feels a bit weird to says "falls back to top-level", when _makeEntryWithBackends is called with no topLevel parameter. I think part of it is me not understanding the test, I guess test is fine but its just naming or structure maybe

File: extensions/replication/tasks/ReplicateObject.js (hunk 12 above)
Comment (SylvainSenechal) [reply extensions/replication/tasks/ReplicateObject.js:271]: Ok I meant here we also accept len 0, i was thinking enforcing either 1 or 2, without more than 2 or lesser than one but its nit whatever its fine

**Review by SylvainSenechal — APPROVED** [review]
Not much to comment, it's really nice that you were able to remove callbacks and turn some functions asynchronous.
This is hard to review, I think it's ok, I left 1 or 2 comments in Arsenal as I was reviewing here, but they are also not very important comments.
You already know but we will have to test this properly, what worries me the most is not people setting up brand new replication from this new branch, but breaking existing replication flow

File: tests/unit/lib/models/ObjectQueueEntry.spec.js (hunk 20 above)
Comment (maeldonn) [reply tests/unit/lib/models/ObjectQueueEntry.spec.js:66]: "Top-level" refers to replicationInfo.destination / .role (what per-backend entries override). The helper was hiding those behind defaults. I've hoisted them into TOP_LEVEL_DEST / TOP_LEVEL_ROLE constants so the fallback assertions reference them directly. Kept the test names since the wording matches the actual replicationInfo shape.

File: extensions/mongoProcessor/MongoQueueProcessor.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/mongoProcessor/MongoQueueProcessor.js:349]: Confirmed safe even for unupgraded cloudserver: grepped cloudserver and there are zero reads of `replicationInfo.destination` / `.storageClass` / `.storageType` (or the corresponding accessors); the only references are to `backends[].destination` which is unchanged. Combined with the answer in https://github.com/scality/backbeat/pull/2743#discussion_r3396323671 (Backbeat readers/writers ship together, and arsenal's per-backend getters fall back to top-level), no data-model migration is needed.

File: extensions/mongoProcessor/MongoQueueProcessor.js [hunk 21]
```diff
@@ -340,53 +322,33 @@ class MongoQueueProcessor {
...
+            status: 'PENDING',
+            backends,
+            content,
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/mongoProcessor/MongoQueueProcessor.js:348]: Confirmed safe even for unupgraded cloudserver: grepped cloudserver and there are zero reads of `replicationInfo.destination` / `.storageClass` / `.storageType` (or the corresponding accessors); the only references are to `backends[].destination` which is unchanged. Combined with the answer in https://github.com/scality/backbeat/pull/2743#discussion_r3396323671 (Backbeat readers/writers ship together, and arsenal's per-backend getters fall back to top-level), no data-model migration is needed.

File: extensions/replication/tasks/ReplicateObject.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/replication/tasks/ReplicateObject.js:271]: Good catch — the fallback assumed `roles[1]` was defined, which broke legacy single-ARN configs. Fixed: when `roles.length === 1`, `expectedDestRole` now falls back to `roles[0]` (single ARN serves both source and destination, per `ReplicationConfiguration.resolveDestinationRole`).

[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/replication/tasks/ReplicateObject.js:271]: The "undefined !== undefined → silently passes" premise does not hold: the entry-side check at line 236 enforces `entryRoles.length === 2`, so `entryRoles[1]` is always defined. The real issue (the BadRole rejection for single-ARN bucket configs) is addressed in the sibling thread https://github.com/scality/backbeat/pull/2743#discussion_r3415042491.

File: extensions/replication/tasks/MultipleBackendTask.js [hunk 22]
```diff
@@ -85,7 +85,8 @@ class MultipleBackendTask extends ReplicateObject {
             .then(data => {
                 const replicationEnabled = data.ReplicationConfiguration.Rules
                     .some(rule => rule.Status === 'Enabled' &&
-                        entry.getObjectKey().startsWith(rule.Prefix));
+                        entry.getObjectKey().startsWith(
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/replication/tasks/MultipleBackendTask.js:88]: Already addressed — `MultipleBackendTask._setupRolesOnce` passes the backend at line 59: `entry.getReplicationRoles(entry.getReplicationBackend())`, mirroring `ReplicateObject._setupRolesOnce` line 231. (The referenced `getBackendKey()` method does not exist; the line numbers in the comment look stale.)

File: extensions/replication/tasks/MultipleBackendTask.js [hunk 23]
```diff
@@ -317,7 +318,7 @@ class MultipleBackendTask extends ReplicateObject {
         const command = new MultipleBackendAbortMPUCommand({
             Bucket: sourceEntry.getBucket(),
             Key: sourceEntry.getObjectKey(),
-            StorageType: sourceEntry.getReplicationStorageType(),
+            StorageType: this._getReplicationEndpointType(),
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/replication/tasks/MultipleBackendTask.js:322]: Already addressed — both call sites now pass the backend object: line 1221 reads `refreshedEntry.getReplicationSiteStatus(sourceEntry.getReplicationBackend())` and line 1242 reads `sourceEntry.getReplicationSiteDataStoreVersionId(sourceEntry.getReplicationBackend())`. (The referenced `getBackendKey()` does not exist; the line numbers in the comment are stale.)

File: extensions/replication/queueProcessor/QueueProcessor.js [hunk 24]
```diff
@@ -865,41 +864,75 @@ class QueueProcessor extends EventEmitter {
...
+        if (!(sourceEntry instanceof ObjectQueueEntry)) {
+            logSkip();
+            return;
+        }
+
+        const pendingBackends = sourceEntry.getReplicationBackends()
+            .filter(b => b.status === 'PENDING' && b.site === this.site);
+
+        if (pendingBackends.length === 0) {
+              logSkip();
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/replication/queueProcessor/QueueProcessor.js]: Already fixed — the lines now use the file standard 4-space indent (12 spaces at this nesting level), verified with `cat -A`.

File: lib/api/BackbeatAPI.js [hunk 25]
```diff
@@ -430,9 +430,8 @@ class BackbeatAPI {
                 const bucket = entry.getBucket();
                 const objectKey = entry.getObjectKey();
                 const mdObj = JSON.parse(res.Body);
-                const queueEntry =
-                    new ObjectQueueEntry(bucket, objectKey, mdObj);
-                queueEntry.setSite(entry.getSite());
+                const queueEntry = new ObjectQueueEntry(bucket, objectKey, mdObj)
+                    .setReplicationBackend({ site: entry.getSite() });
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply lib/api/BackbeatAPI.js]: Good catch — confirmed it is a real regression, not intended. Failures in Redis are tracked per `(bucket, key, version, role)` (`ObjectFailureEntry` encodes the role into the sorted-set member), so each retry candidate is already keyed by role. Pre-PR there was at most one backend per site so `_findBackend({site})` was unambiguous; with multi-destination it now silently matches whichever backend is first in `backends[]`.

Fixed in `_getObjectQueueEntry`: we now look up the matching backend from the fetched MD using `(site, role)` and stamp `{site, role, destination}` so `_findBackend` resolves the exact backend the failure refers to. Added a unit test in `tests/unit/api/BackbeatAPI.spec.js` that feeds two backends sharing `site=sf` but differing in destination/role, and asserts the queueEntry stamps the second backend (matching the failure role) rather than defaulting to the first.

The test gap that let this slip: `_pushToCRRRetryKafkaTopics` is already tested for the multi-destination payload, but no test wired `_getObjectQueueEntry` → `_pushToCRRRetryKafkaTopics` with multi-destination MD, so the seam on line 434 was unexercised.

File: extensions/replication/queueProcessor/QueueProcessor.js [hunk 26]
```diff
@@ -865,41 +864,91 @@ class QueueProcessor extends EventEmitter {
...
         }
-        if (task) {
-            this.logger.debug('replication entry is being pushed', { entry: sourceEntry.getLogInfo() });
-            return this.taskScheduler.push({ task, entry: sourceEntry, kafkaEntry }, done);
+        if (firstError) {
+            throw firstError;
         }
-        this.logger.debug('skip replication entry', { entry: sourceEntry.getLogInfo() });
-        return process.nextTick(done);
     }
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/replication/queueProcessor/QueueProcessor.js:953]: Good catch — the comment misattributed the safety to the filter. Reworded to make clear that re-dispatch on retry is safe because each task refreshes live metadata and short-circuits already-completed backends, not because the Kafka filter mutates.

File: lib/models/ObjectQueueEntry.js [hunk 27]
```diff
@@ -179,58 +230,60 @@ class ObjectQueueEntry extends ObjectMD {
...
             .setReplicationStatus('PENDING');
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply lib/models/ObjectQueueEntry.js:273]: Acknowledged — intentional. Routing now keys off `backends[]` (narrowed to `[matched]` on line 271), and the top-level `storageClass` is unused by downstream consumers (confirmed in https://github.com/scality/backbeat/pull/2743#discussion_r3395972484 and the cloudserver grep in the sibling thread https://github.com/scality/backbeat/pull/2743#discussion_r3437282107). No need to mutate the top-level field on retry entries.

File: tests/unit/lib/models/ObjectQueueEntry.spec.js (hunk 19 above)
Comment (maeldonn) [reply tests/unit/lib/models/ObjectQueueEntry.spec.js:62]: Thanks for the pointer — agreed, ARNs here are bucket/role identifiers, not object keys, so the colon-in-key parsing edge case does not apply. Keeping it in mind for any future arsenal work that does cross object-key ARN territory.

File: lib/api/BackbeatAPI.js [hunk 28]
```diff
@@ -430,9 +430,17 @@ class BackbeatAPI {
...
                 const objectKey = entry.getObjectKey();
                 const mdObj = JSON.parse(res.Body);
-                const queueEntry =
-                    new ObjectQueueEntry(bucket, objectKey, mdObj);
-                queueEntry.setSite(entry.getSite());
+                const site = entry.getSite();
+                const failureRole = entry.getReplicationRoles() || undefined;
+                const mdBackend = (mdObj.replicationInfo?.backends || [])
+                    .find(b => b.site === site &&
+                        (!failureRole || b.role === failureRole));
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply lib/api/BackbeatAPI.js]: You are right — confirmed by checking `UpdateReplicationStatus._pushFailedEntry` (line 105): it stores `getReplicationRoles().split(",")[0]`, which is the **source** role, not the per-backend destination role. My match against `b.role` would never succeed, and `_findBackend` would then build an empty `backends[]` on retry — strictly worse than the prior "first-match-by-site" behavior. Reverting the change and the associated test.

The actual disambiguation needs a Redis-schema change (store per-backend destination role in the failure record, update the API response and filter accordingly). I will track that as a follow-up; for this PR the prior "match first backend by site" is preserved.

File: lib/api/BackbeatAPI.js (hunk 25 above)
[the remark this answers is not in the document]
Comment (maeldonn) [reply lib/api/BackbeatAPI.js]: Update: revert in https://github.com/scality/backbeat/pull/2743#discussion_r3437669264 — the role stored in Redis is the source role (set by `_pushFailedEntry`), not the per-backend destination role, so the match never succeeded. Tracked as a follow-up requiring a Redis-schema change.

File: extensions/replication/tasks/UpdateReplicationStatus.js [hunk 29]
```diff
@@ -238,12 +240,11 @@ class UpdateReplicationStatus extends BackbeatTask {
...
         } else if (newStatus === 'PENDING') {
-            entry = refreshedEntry.toPendingEntry(site);
+            entry = refreshedEntry.toPendingEntry(backend);
         }
-        const versionId =
-            sourceEntry.getReplicationSiteDataStoreVersionId(site);
-        entry.setReplicationSiteDataStoreVersionId(site, versionId);
-        entry.setSite(site);
+        const versionId = sourceEntry.getReplicationSiteDataStoreVersionId(backend);
+        entry.setReplicationSiteDataStoreVersionId(backend, versionId);
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/replication/tasks/UpdateReplicationStatus.js:246]: Does not apply here: line 209 reads `newStatus` from `sourceEntry`, not `refreshedEntry`, and `_checkStatus(sourceEntry)` at line 438 has already validated that the same `sourceEntry` + `backend` returns a status in `["COMPLETED","FAILED","PENDING"]`. Since `sourceEntry` is not mutated between those two calls, `newStatus` is guaranteed to take one of the three branches.

File: extensions/replication/queueProcessor/QueueProcessor.js [hunk 30]
```diff
@@ -865,41 +864,92 @@ class QueueProcessor extends EventEmitter {
...
+        // Sequential loop over pending backends. Each backend is attempted
+        // independently so a failure on one does not skip the others; the
+        // first error is re-thrown at the end so the Kafka entry is retried.
+        // On retry, the same backends are re-dispatched: the task itself
+        // refreshes the live metadata and short-circuits backends that have
+        // since completed.
+        let firstError = null;
+        for (const backend of pendingBackends) {
+            const task = new (isCloud ? MultipleBackendTask : ReplicateObject)(this);
+            const perBackendEntry = sourceEntry.clone().setReplicationBackend(backend);
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/replication/queueProcessor/QueueProcessor.js:924]: Confirmed intentional — the sequential design is documented in the comment at lines 915-920 and was the approach approved in https://github.com/scality/backbeat/pull/2743#discussion_r3395990723 (per-backend try/catch, first error re-thrown so the Kafka entry is retried). Per-task metadata refresh handles the clone-isolation case you describe.

File: extensions/replication/tasks/ReplicateObject.js [hunk 31]
```diff
@@ -287,18 +289,45 @@ class ReplicateObject extends BackbeatTask {
...
-            if (roles[1] !== entryRoles[1]) {
+            // Pick the rule for this specific backend (multi-destination
+            // configs may share a StorageClass with distinct Bucket /
+            // Account), then derive the expected destination role from
+            // its Account; fall back to literal role[1] for legacy
+            // configs without Account.
+            const entryDestination = entry.getDestination();
+            const entryRoleAccount = entry.getRole()?.split(':')[4];
+            const matchingRule = data.ReplicationConfiguration.Rules.find(rule => {
+                const prefix = rule.Filter?.Prefix ?? rule.Prefix ?? '';
```
[the remark this answers is not in the document]
Comment (maeldonn) [reply extensions/replication/tasks/ReplicateObject.js:300]: The fallback chain is `matchingRule+Account → roles[1] (when length 2) → roles[0] (single-ARN)`. With no matching rule and a normal 2-role config we land in `roles[1]` (the bucket destination role), which is exactly what the entry was built from in `MongoQueueProcessor` via `resolveDestinationRole(bucketRepInfo.role)` — matches by construction. The `roles[0]` branch is only reached for legacy single-ARN configs, where the same ARN serves as both source and destination so it still matches `entryRoles[1]`. So neither case rejects valid entries.
